drift.py

```diff
@@ -1,146 +1,90 @@
-            if len(spec[verToken:]) > 0:
-                result.version = spec[verToken+1:]
-            if out.returncode == SUCCESS:
-                concrete_spec = json.loads(out.stdout)["spec"]
-                # In the case that the concretized spec hash changes
-                # we want to record the commit where this happens.
-                if self.last[spec] != "" and self.last[spec][0][name]["full_hash"] != concrete_spec[0][name]["full_hash"]:
-                    # Check if the package version is different.
-                    if self.last[spec][0][name]["version"] != concrete_spec[0][name]["version"]:
-                        result.tags += ["pkg-updated"]
-                    # +---------------------+
-                    # |     Dependencies    |
-                    # +---------------------+
-                    result.tags += set(self.tag_deps(concrete_spec, spec, name))
-                    # +---------------------+
-                    # |      Variants       |
-                    # +---------------------+
-                    result.tags += set(self.tag_variants(concrete_spec, spec, name))
-                    # Send Inflection Point
-                    send(result)
-                # Save concrete spec hash as last hash.
-                self.last[spec] = concrete_spec
-            else:
-                # If the spec doesn't concretize properly we also want
-                # to record the commit at which this occurred.
-                result.tags += ["concretizaton-failed"]
+        for abstract_spec in self.specs:
+            # Separate out spec version and name.
+            name = abstract_spec
+            version = ""
+            if "@" in abstract_spec:
+                name, version = abstract_spec.split("@", 1)
+            # Check that version exists in package.py file.
+            out = run(f"spack versions --safe-only {abstract_spec}")
+            # Don't attempt to concretize if the version doesn't yet exist.
+            if version in out.stdout:
+                out = run(f"spack spec --yaml {abstract_spec}")
+                # If concretization successful check the resulting concrete specs.
+                if out.returncode == SUCCESS:
+                    concrete_spec = spack.spec.Spec().from_yaml(out.stdout)
+                    # If the spec is the same move on.
+                    if abstract_spec not in self.last:
+                        self.last[abstract_spec] = concrete_spec
+                    if self.last[abstract_spec] == concrete_spec:
+                        continue
+                    # If the spec is different check what is different and save as tags.
+                    diff = compare_specs(
+                        self.last[abstract_spec],
+                        concrete_spec,
+                        colorful=False)
+                    # Flatten list to tags with context.
+                    tags = ["%s:%s" %(x[0],x[1]) for x in diff['b_not_a']]
```

Comment:
I think you should save both specs and both sides of the diff -- it will give us more information and we can come back and compare them differently later if we want.  `a_not_b` may actually be relevant (and interesting to learn from) here.

Reply:
@vsoch do you want to meetup to talk about how we should refactor the drift server to support both sides of the diff?

@tgamblin we were actually planning on saving the concrete spec at the inflection point during the build process not during this analysis process just in case anything changes from analysis time to build time we know we're saving the concrete spec actually used when the build failed/succeeded. Do you think that is sufficient or should we instead save the specs here?

Reply:
if you know the concrete spec before and after the inflection point, then I do not think this matters -- though I generally think it is best to save the data as soon as you know it.  You figure out inflection points and decide whether to build here right?  In that case it seems like you want to also save the concrete specs here, so that you could run through all the builds again.  I guess you can do that anyway if you just have the commit and the abstract spec.

Reply:
> You figure out inflection points and decide whether to build here right?

Yes? Here is where we tag an inflection point and upload it to the drift-server but this isn't where we're doing the builds.  That comes later with another application drift-builder which queries the server for all of the inflection points for a given abstract spec and then loops through those points building at each and reporting back the concretized spec built and whether it built successfully or not.